zebra-network/src/peer_set/set.rs

```diff
@@ -423,6 +423,12 @@ where
                     .ready_services
                     .get_index_mut(index)
                     .expect("preselected index must be valid");
+
+                // `PeerSet` calls `poll_ready` multiple times on its `Client` services.
+                // `Buffer`ed polls can fill up their Buffer reservations, causing hangs.
+                // The next `assert` will make sure the service type is never `Buffer`. #1593
+                assert!(!format!("{:?}", type_of(&service)).contains("Buffer"));
```

Comment:
```suggestion
                assert!(!format!("{:?}", type_of(&service)).contains("Buffer"),
                             "Clients must not use tower::Buffer, because PeerSet uses poll_ready multiple times before each call, which can cause hangs with buffers" );
```